drivers/codec/wm8960.py

```diff
@@ -0,0 +1,783 @@
+            )
+
+        elif module == module_mic_bias:
+
+            regs[_POWER1] = (
+                _POWER1_MICB_MASK,
+                (is_enabled << _POWER1_MICB_SHIFT),
+            )
+
+        elif module == module_speaker:
+
+            regs[_POWER2] = (
+                _POWER2_SPKL_MASK | _POWER2_SPKR_MASK,
+                (_POWER2_SPKL_MASK | _POWER2_SPKR_MASK) * is_enabled,
+            )
+            regs[_CLASSD1] = 0xF7
+
+        elif module == module_omix:
+
+            regs[_POWER3] = (
+                _POWER3_LOMIX_MASK | _POWER3_ROMIX_MASK,
+                (_POWER3_LOMIX_MASK | _POWER3_ROMIX_MASK) * is_enabled,
+            )
+
+        elif module == module_mono_out:
+
+            regs[_MONOMIX1] = regs[_MONOMIX2] = is_enabled << 7
+            regs[_MONO] = is_enabled << 6
+
+        else:
+            raise ValueError("Invalid module")
+
+    def enable_module(self, module):
+        self.set_module(module, True)
+
+    def disable_module(self, module):
+        self.set_module(module, False)
+
+    def set_data_route(self, route):
+        regs = self.regs
+        if route == route_bypass:
+            # Bypass means from line-in to HP
+            # Left LINPUT3 to left output mixer, LINPUT3 left output mixer volume = 0dB
+            # Right RINPUT3 to right output mixer, RINPUT3 right output mixer volume = 0dB
+            regs[_LOUTMIX, _ROUTMIX] = 0x80
+
+        elif route == route_playback:
+            # Data route I2S_IN-> DAC-> HP
+            #
+            # Left DAC to left output mixer, LINPUT3 left output mixer volume = 0dB
+            # Right DAC to right output mixer, RINPUT3 right output mixer volume = 0dB
+            regs[_LOUTMIX, _ROUTMIX] = 0x100
+            regs[_POWER3] = 0x0C
+            # Set power for DAC
+            self.set_module(module_DAC, True)
+            self.set_module(module_omix, True)
+            self.set_module(module_line_out, True)
+
+        elif route == route_playback_record:
+            #
+            # Left DAC to left output mixer, LINPUT3 left output mixer volume = 0dB
+            # Right DAC to right output mixer, RINPUT3 right output mixer volume = 0dB
+            regs[_LOUTMIX, _ROUTMIX] = 0x100
+            regs[_POWER3] = 0x3C
+            self.set_module(module_DAC, True)
+            self.set_module(module_ADC, True)
+            self.set_module(module_line_in, True)
+            self.set_module(module_omix, True)
+            self.set_module(module_line_out, True)
+
+        elif route == route_record:
+            # LINE_IN->ADC->I2S_OUT
+            # Left and right input boost, LIN3BOOST and RIN3BOOST = 0dB
+            regs[_POWER3] = 0x30
+            # Power up ADC and AIN
+            self.set_module(module_line_in, True)
+            self.set_module(module_ADC, True)
+
+        else:
+            raise ValueError("Invalid route")
+
+    def set_left_input(self, input):
```

Comment:
It could be possible to simplify this method by having the input selection values more complex types.

Eg:
```python
INPUT_CLOSED = None
INPUT_MIC1 = (0x138, 0x117)
INPUT_MIC2 = (0x178, 0x117)
INPUT_MIC3 = (0x1b8, 0x117)
INPUT_LINE2 = (0, 0xe)
INPUT_LINE3 = (0, 0x70)

def set_left_input(self, input):
    if input is None:
        regs[_POWER1] = (_POWER1_AINL_MASK | _POWER1_ADCL_MASK, 0)
    elif input[0] == 0:
        regs[_POWER1] = (0, _POWER1_AINL_MASK | _POWER1_ADCL_MASK)
        regs[_INBMIX1] = input
    else:
        regs[_POWER1] = (0, _POWER1_AINL_MASK | _POWER1_ADCL_MASK | _POWER1_MICB_MASK)
        regs[_LINPATH] = input[0]
        regs[_LINVOL] = input[1]
```
And `set_right_input()` would also be simplified, and can use the same constants.

Reply:
That looks good, but I would like to do it in the same way as for the volumes, with an internal dictionary of the value pairs. Then INPUT_XXXX being method arguments could stay an integer constants, and the value can be checked for a valid value.

Reply:
Done and pushed.